Flow allocator with bandwidth reservation

autrimpo opened the pull request.

autrimpo commented:
Hi @vmaffione, do you want me to do some refactoring and cleanup before doing a code review?

vmaffione commented:
Yes if you want your code to be merged. You can have a commit introducing the changes affecting existing functionalities, and another one that actually introduces the bwres mechanism. Also you need to fix integration tests (see "Details" on the travis info above).

autrimpo commented:
Okay, I'll start working on it.

autrimpo commented:
Hopefully this looks okay, let me know if you want any more changes.

vmaffione commented:
Thanks, I'll let you know asap.

vmaffione commented on user/libs/ker-numtables.c:
```diff
@@ -0,0 +1 @@
+../../common/ker-numtables.c
```
Remove this file, as it is generated by `./configure`

vmaffione replied:
You have not removed these.
Please use `git rm user/libs/ker-numtables.c`.

vmaffione commented on user/libs/utils.c:
```diff
@@ -0,0 +1 @@
+../../common/utils.c
```
Remove this file, as it is generated by `./configure`

vmaffione replied:
Same as above.

vmaffione commented on README.md:
```diff
@@ -1074,6 +1077,16 @@ of a normal IPCP process:
 | flowalloc           | local             | initial-a          | Initial value for the DTCP A timer. |
 | flowalloc           | local             | initial-credit     | Initial size of the DTCP flow control window (in PDUs). |
 | flowalloc           | local             | max-cwq-len        | Maximum size of the DTCP closed window queue (in PDUs). |
+| flowalloc           | bw-res            | force-flow-control | If false, flow control is used only with reliable flows. If true, flow control is always used. |
```
For common parameters, set `*` in the `Policy` fields to avoid duplication.

vmaffione commented on demo/libdemo.py:
```diff
@@ -515,14 +515,18 @@ def compute_shim_ipcps(self, vm, mac2ifname = True):
                 vars_dict['ifname'] = port['veth']
                 outs += 'PORT%(idx)s=%(ifname)s\n' % vars_dict
             outs += '$SUDO ip link set $PORT%(idx)s up\n' % vars_dict
+            outs += '$SUDO ethtool -s $PORT%(idx)s speed %(shimspeed)s\n' % vars_dict
```
Shim speed is normally zero, which means you don't want to set a speed. You should not execute this command if shim speed is zero.
In addition to that, setting the speed in that way may have effect with physical NICs (although it may lead to link negotiation error). It will have no effect on VMs, which is what the demonstrator is meant for, because QEMU does not emulate link speed. So why do you need this?

vmaffione commented on user/uipcps/uipcp-container.c:
```diff
@@ -1582,3 +1596,36 @@ uipcp_update(struct uipcps *uipcps, struct rl_kmsg_ipcp_update *upd)
+        return -1;
+    }
+
+    struct ethtool_cmd edata;
```
For now we use the older C standard, which means that variables must be declared at the beginning of the function.

vmaffione commented on user/uipcps/uipcp-container.c:
```diff
@@ -1582,3 +1596,36 @@ uipcp_update(struct uipcps *uipcps, struct rl_kmsg_ipcp_update *upd)
+    struct ethtool_cmd edata;
+    edata.cmd = ETHTOOL_GSET;
+
+    struct ifreq ifr;
```
It's good practice to zero-init the `ifr` structure before starting to fill it.

vmaffione commented on user/uipcps/uipcp-normal-lfdb.hpp:
```diff
@@ -61,11 +61,18 @@ using NodeId = std::string;
  * Free Alternate algorithm. */
 struct LFDB {
     struct Edge {
+        NodeId from;
```
The `from` field is redundant, as this struct is meant to be used combined in this way:
````
unordered_map<NodeId, vector<Edge>> g;
````
and so `g[x][i].from == x for each x and for each i`.
Can you please remove it?

vmaffione commented on user/uipcps/uipcp-normal-lfdb.cpp:
```diff
@@ -254,6 +255,132 @@ LFDB::compute_next_hops(const NodeId &local_node)
+        for (const Edge &edge : edges) {
+            if (pred.find(edge.to) == pred.end() &&
+                (pred.find(cur) == pred.end() ||
+                 edge.to != pred.find(cur)->second.from) &&
```
this would be `edge.to != pred.find(cur)->first`

vmaffione commented on user/uipcps/uipcp-normal-lfdb.cpp:
```diff
@@ -254,6 +255,132 @@ LFDB::compute_next_hops(const NodeId &local_node)
+    if (pred.find(dest_node) != pred.end()) {
+        path.insert(path.begin(), dest_node);
+        for (auto e = pred.find(dest_node); e != pred.end();
+             e      = pred.find(e->second.from)) {
```
this would be `pred.find(e->first)`

vmaffione commented on user/uipcps/uipcp-normal-lfdb.cpp:
```diff
@@ -254,6 +255,132 @@ LFDB::compute_next_hops(const NodeId &local_node)
+                (pred.find(cur) == pred.end() ||
+                 edge.to != pred.find(cur)->second.from) &&
+                edge.capacity >= req_flow) {
+                pred.emplace(edge.to, edge);
```
I guess here you would need to use something like
````
pred.emplace(edge.to, Edge(cur, /*cost*/)).
````

autrimpo commented:
I've fixed the issues you've mentioned.
Sorry for the delay but the graph algorithm was getting stuck in a loop, as some of the semantics have changed with the removeal of `.from`, so I had to write the algorithm down and work it through step by step to figure out how to change it.

The pipeline fails, which seems to be related to [this bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53119). Would it be possible to upgrade the gcc version used in the pipeline?

vmaffione commented:
Thanks a lot. Take your time.
Unfortunately we don't have much control on the travis build environment. However, this is not a problem, as the `{0}` initializer is not in C99 (that we are using). Can you please replace that with `memset(&ifr, 0, sizeof(ifr));`?

vmaffione commented on user/uipcps/uipcp-normal-flow-alloc.cpp:
```diff
@@ -721,6 +728,940 @@ LocalFlowAllocator::dump_memtrack(std::stringstream &ss) const
+class BwResFlowAllocator : public LocalFlowAllocator {
+    class Replica : public CeftReplica {
+        struct Command {
+            char flow_id[127];
```
Instances of this struct will be stored in the raft log as an array.
It makes sense to lay out the fields in such a way to avoid padding, i.e. ensuring natural boundaries.
This means that char arrays should be multiple of 8, or better 16.
One of the fields can be `31` simply because the 32th byte contains the `opcode` field, which must follow immediately after..

vmaffione commented on user/uipcps/uipcp-normal-flow-alloc.cpp:
```diff
@@ -721,6 +728,940 @@ LocalFlowAllocator::dump_memtrack(std::stringstream &ss) const
+            char from[31];
+            char to[31];
+            uint32_t bw;
+            uint8_t opcode;
```
move this right after `to[31]`

autrimpo commented:
I'll submit another PR after this one adding all the config-generated files to gitignore, is that okay?

vmaffione commented:
Yes, feel free to open that whenever you want. That one can go in immediatly.

vmaffione commented on user/uipcps/uipcp-normal-lfdb.hpp:
```diff
@@ -63,9 +63,14 @@ struct LFDB {
-        Edge(Edge &&) = default;
+        unsigned int capacity;
+
+        Edge(const NodeId &to_, unsigned int cost_) : to(to_), cost(cost_){};
```
small suggestion: move this constructor after the following one, and use constructor delegation:
````
: Edge(to_, cost_, 0) {}
````
This also makes sure that `capacity` is always initialized

autrimpo commented:
Pushed an amended commit fixing the constructor issue.

vmaffione commented on user/uipcps/lfdb-bw-test.cpp:
```diff
@@ -0,0 +1,197 @@
+     * their available bandwidth. An item in list of reachability tests is a
+     * tuple of integers; the first integer refers to the source node, the
+     * second to the destination node and the third is the requested bandwidth
+     * for the flow. */
```
It's not clear what the semantic of the list of reachability tests is. In the first test vector, for example, are the four flow allocations occuring one after the other, on the same graph, i.e. taking into account what was allocated in the previous flow allocation?

vmaffione commented on user/uipcps/lfdb-bw-test.cpp:
```diff
@@ -0,0 +1,197 @@
+    auto usage = []() {
+        std::cout << "lfdb-test -n SIZE\n"
+                     "          -v be verbose\n"
+                     "          -h show this help and exit\n";
```
The 'n' option is not used. Please remove it.

vmaffione commented on user/uipcps/lfdb-bw-test.cpp:
```diff
@@ -0,0 +1,197 @@
+    int opt;
+    n++;
+
+    while ((opt = getopt(argc, argv, "hnv:")) != -1) {
```
Remove `n`.

vmaffione commented on user/uipcps/lfdb-bw-test.cpp:
```diff
@@ -0,0 +1,197 @@
+            verbosity++;
+            break;
+
+        case 'n':
```
remove `n`

vmaffione commented on user/uipcps/lfdb-bw-test.cpp:
```diff
@@ -0,0 +1,197 @@
+        auto start = std::chrono::system_clock::now();
+
+        for (const auto &rtest : reachability_tests) {
+            lfdb.reachable(std::get<0>(rtest), std::get<1>(rtest),
```
You are not checking anything here. You should check that `reachable()` returns true or false depending on the test case. You should store the expected value (`true` or `false` in the ReachabilityTest tuple).

vmaffione commented on user/uipcps/lfdb-bw-test.cpp:
```diff
@@ -0,0 +1,197 @@
+
+    std::cout << "Flow " << req_flow << " from " << cur << " to " << dst
+              << std::endl;
+    find_flow_path(cur, dst, req_flow);
```
Grab the return value and return `true` if a path was found, `false` otherwise.

vmaffione commented on user/uipcps/lfdb-bw-test.cpp:
```diff
@@ -0,0 +1,197 @@
+};
+
+/* Returns true if the routing tables are able to route a packet from
+ * 'src_node' to 'dst_node' with exactly 'n' hops. */
```
This comment should be updated to reflect what is happening.

vmaffione commented on user/uipcps/uipcp-container.c:
```diff
@@ -1582,3 +1596,42 @@ uipcp_update(struct uipcps *uipcps, struct rl_kmsg_ipcp_update *upd)
+
+    int ret;
+    int speed;
+
```
You must check if `if_name` is NULL, and return `-1` in that case, because the config get operation failed.
Also, you should check if `if_name` is the empty string, which means that the `netdev` param is not configured yet. If this is the case, there is nothing to do, and you can just `return 0` immediately.

vmaffione commented on user/uipcps/uipcp-container.c:
```diff
@@ -1582,3 +1596,42 @@ uipcp_update(struct uipcps *uipcps, struct rl_kmsg_ipcp_update *upd)
+        UPW(uipcp, "uipcp_get_if_speed() failed [%s]\n", strerror(errno));
+        return -1;
+    }
+
```
add `memset(&edata, 0, sizeof(edata));`
to make sure `edata` has a known value and does not leak data into the kernel.

vmaffione commented on user/uipcps/uipcp-container.c:
```diff
@@ -1582,3 +1596,42 @@ uipcp_update(struct uipcps *uipcps, struct rl_kmsg_ipcp_update *upd)
+    ret = ioctl(skfd, SIOCETHTOOL, &ifr);
+    if (ret < 0) {
+        UPW(uipcp, "uipcp_get_if_speed() failed: ioctl on %s: [%s]\n", if_name,
+            strerror(errno));
```
`close(skfd)` here, otherwise you have a resource leak.

vmaffione commented on user/uipcps/uipcp-container.c:
```diff
@@ -1582,3 +1596,42 @@ uipcp_update(struct uipcps *uipcps, struct rl_kmsg_ipcp_update *upd)
+
+    speed       = ethtool_cmd_speed(&edata);
+    uipcp->if_speed = speed == -1 ? 0 : speed * 1000 * 1000; // Mbps -> bps
+
```
`close(skfd)` here, otherwise you have a resource leak.

vmaffione commented on user/uipcps/uipcp-container.h:
```diff
@@ -260,6 +260,9 @@ struct uipcp {
     /* Siblings. */
     struct list_head node;
+
+    /* Interface speeds */
```
`Interface speed.`

vmaffione commented on user/uipcps/uipcp-normal-flow-alloc.cpp:
```diff
@@ -22,11 +22,17 @@
 
 using namespace std;
 
+using NodeId = std::string;
```
We should not define the same symbol twice, risking inconsistencies.
Just move this statement from `user/uipcps/uipcp-normal-lfdb.hpp` to `user/uipcps/uipcp-normal.hpp`, so that it is visible to all the modules.

autrimpo commented on CMakeLists.txt:
```diff
@@ -2,6 +2,8 @@ cmake_minimum_required(VERSION 2.8)
 
 project(rlite)
 
+set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
```
The generated compile-commands file is useful for various language servers such as clangd or ccls to provide code completion, jumping to definition, finding references etc. across different editors.
See [the Language Server homepage](https://langserver.org/) for more information.
Most C/C++ servers depend on this database file for correct functionality.

vmaffione commented on user/uipcps/uipcp-normal-flow-alloc.cpp:
```diff
@@ -721,6 +728,940 @@ LocalFlowAllocator::dump_memtrack(std::stringstream &ss) const
+            m->invoke_id = rm->invoke_id;
+
+            auto cbuf  = std::unique_ptr<char[]>(new char[sizeof(Command)]);
+            Command *c = reinterpret_cast<Command *>(cbuf.get());
```
You must `memset(c, 0, sizeof(*c))` to ensure the struct is properly initialized

autrimpo commented:
Thank you, are you done with reviewing? i'll wait until you go through everything before making changes.

vmaffione commented on user/uipcps/uipcp-normal-flow-alloc.cpp:
```diff
@@ -721,6 +728,940 @@ LocalFlowAllocator::dump_memtrack(std::stringstream &ss) const
+        remote_addr, rm->result ? 1 : 0, &freq->flowcfg);
+}
+
+int
```
This looks the very same code as the base class. Why replicating?

vmaffione replied:
You can reuse the code and `#if 0` the `expected_src_addr` check.

autrimpo replied:
Wouldn't that remove the check from the base class as well?

vmaffione replied:
Yes, but, I do prefer to temporarily remove that check so that you reuse the code, and later look for a way to restore the check only where needed.

vmaffione commented on user/uipcps/uipcp-normal-flow-alloc.cpp:
```diff
@@ -721,6 +728,940 @@ LocalFlowAllocator::dump_memtrack(std::stringstream &ss) const
+
+/* (4) Initiator FA <-- Slave FA : M_CREATE_R */
+int
+BwResFlowAllocator::flows_handler_create_r(const CDAPMessage *rm, int invoke_id)
```
This looks the same as the base class. Can you drop it? The base class method will be called.

vmaffione commented:
I am done for now, but since the change is complex and I already dropped many comments, I would like to give a further pass once you address my comments.

A general question: I assume (and it looks like) the implementation follows the guidelines that we agreed on the "Bw allocation scheme" google doc. Can you confirm?

One big thing that is missing in this pull request is an integration test, maybe something similar to 
`tests/integration/0027-ceft-addralloc.sh`. Such a test would ensure that we never break your code in the future. With so much complexity, it's very likely that a future unrelated change will break it (it happened multiple times in the past, before I introduced the integration tests framework).
Even before that, it would ensure that none of the changes you will make to address my comments will break it. Maybe you can just prepare a test that reproduces the scenario that you setup manually to test your code (I guess it's a 4 or 5 nodes scenario).

autrimpo commented:
I see, in that case I will start working on the issues.

Yes, the implementation should be consistent with the final version of the scheme we agreed on in the google doc.

The test scenario used 5 nodes, I should be able to find the demonstrator config as well as possibly some scripts I used to test the implementation. I will begin by creating an integration test.

vmaffione commented:
Excellent, thanks. Feel free to add the demonstrator config file to the example if you think it is in shape.

autrimpo commented on user/uipcps/uipcp-normal-lfdb.cpp:
```diff
@@ -254,6 +255,132 @@ LFDB::compute_next_hops(const NodeId &local_node)
+        for (const Edge &edge : edges) {
+            if (pred.find(edge.to) == pred.end() &&
+                (pred.find(cur) == pred.end() ||
+                 edge.to != pred.find(cur)->second.to) &&
```
Indeed, you are right.

autrimpo commented:
Hi Vincenzo, I hope you're doing well.

Sorry for the long delay; I've started working on the PR again. I've resolved all the comments where I worked in your changes, except the one with `#if 0` - I'm not sure I understood you correctly.
I will try to create an integration test asap and after that push my commits.

vmaffione commented:
Hi,
  Sounds good! Please also merge the latest master in your branch. There have been some minor changes, including a bug in your unit test (triggered by a `./configure --debug` build).

autrimpo commented:
I've pushed a partial integration test, based on `0027` and `0028`. I know it's not possible to set the interface speed on `veth` interfaces with `ethtool`, but they do report 10Gbps when created and checked manually. As you can see, `dif-rib-show` reports 0 for all the lower flows. Any clues why that might be? I've checked in the demonstrator with qemu and the speeds are reported correctly there.

(Ignore the build error, I'll look into that.)

autrimpo commented:
I can't reproduce the compilation error with local gcc 5.5.0 (nor the system 9.2.0). I thought rebasing onto current master would help but no luck. What I find curious is that the reconfigure function is practically identical with the `CentralizedFaultTolerantDFT::reconfigure()` in master except policy parameter prefixes - and master builds fine.

vmaffione commented:
> I've pushed a partial integration test, based on `0027` and `0028`. I know it's not possible to set the interface speed on `veth` interfaces with `ethtool`, but they do report 10Gbps when created and checked manually. As you can see, `dif-rib-show` reports 0 for all the lower flows. Any clues why that might be? I've checked in the demonstrator with qemu and the speeds are reported correctly there.
> 
> (Ignore the build error, I'll look into that.)

With QEMU interfaces are virtio-net or e1000, and not veth.
I don't know why you get 0 as interface speed. Have you tried to log the return value of `ethtool_cmd_speed()` to see how many times it is called, and what returns on each call? Maybe it's called too early. Have you tried to write a stand-alone program to test the `uipcp_get_if_speed` and see what happens with veths? In any case looking here 
https://stackoverflow.com/questions/2872058/get-link-speed-programmatically
there are some minor differences with your code, and you may want to look at those. It is also possible to read from `/sys/class/net/eth0/speed`.

vmaffione commented:
> I can't reproduce the compilation error with local gcc 5.5.0 (nor the system 9.2.0). I thought rebasing onto current master would help but no luck. What I find curious is that the reconfigure function is practically identical with the `CentralizedFaultTolerantDFT::reconfigure()` in master except policy parameter prefixes - and master builds fine.

Look at the includes. I think you are missing at least `<algorithm>`. In any case, I agree the error message is really confusing....

autrimpo commented on user/uipcps/uipcp-normal-lower-flows.cpp:
```diff
@@ -681,9 +681,15 @@ BwResRouting::update_local(const string &node_name)
     for (auto f : neigh->flows) {
-        if (f.second->lower_ipcp_id != (rl_ipcp_id_t)RL_IPCP_ID_NONE) {
-            continue;
-        }
```
This was the offending part, seems like I misunderstood your original comment.
Does the current version make sense?
If we don't have a `lower_ipcp_id`, we try to look it up (and possibly return early), else we use the value we already have from the iterator.

vmaffione replied:
I think you should invert the logic:
````
if (f.second->lower_ipcp_id == (rl_ipcp_id_t)RL_IPCP_ID_NONE) {
            continue;
     }
````
The lower ipcp id of a lower flow is not something that you try to look up. Either it's there (since when the NeighFlow object was built), or it will never be there.
Btw, lower ipcp id will be always valid if you don't enable the `reliable-n-flows` parameter.

I'll try to rephrase my original comment. Since you run a loop with the goal of calling `lf->set_bw_total()` and `lf->set_bw_free()`, I think you should check that you actually called that, i.e. check that after the loop bw_free and bw_total are not 0. What happens for a normal IPCP that has a normal IPCP as a lower flow? In this case I think you would have bw_free and bw_total set to 0. How do you behave in this case? If this case is not supported yet, you should probably prevent the installation of the bwres policy for normal IPCPs that are not running over shims.

autrimpo replied:
Yes, this case is not supported yet. I'll look into preventing the policy from being installed.

vmaffione commented on tests/integration/0031-bw-res-flowalloc.sh:
```diff
@@ -0,0 +1,58 @@
+    ip netns exec ${cont} rlite-ctl dif-policy-list bwresdif flowalloc | grep -q "\<bw-res\>"
+    ip netns exec ${cont} rlite-ctl dif-policy-param-list bwresdif flowalloc | grep -q "/mgmt/flowalloc.default-bw = '2000000000'"
+done
+
```
Here you are missing some `rinaperf` tests to check connectivity between A, C and D

autrimpo replied:
Yes, the test isn't done yet, I only pushed the "corrected" code that gets the interface speed so you could look at it.

autrimpo commented:
Am I correct in assuming the Raft logs in `/tmp` should be the same for all the nodes?

autrimpo commented:
I've added a working integration test.

However, it seems that if a node is both the endpoint of a flow and the Raft leader, it gets the `OpcodeFree` message twice for the same flow id, which would double free and crash the node.
For now I've added an additional check when freeing to avoid this, as I haven't managed to track down why exactly this is happening.

Additionally, due to having to test with 10Gbps interfaces, I converted the bandwidth values in the policy to 64 bits else it would overflow. 
What is left unchanged is
- `dtcp_config.bandwidth` in `rlite/common.h`- I wasn't sure if this was implementation defined or spec defined so I left it alone for now
- passing policy parameter values throught `rlite-ctl`

vmaffione commented:
There is a different raft log for each component of each IPCP. For example
/tmp/ceft-aa-18-myname.IPCP
is the log for the address allocator of IPCP 18 (on some host).

Il giorno ven 10 apr 2020 alle ore 13:05 Michal Koutenský <
notifications@github.com> ha scritto:

> Am I correct in assuming the Raft logs in /tmp should be the same for all
> the nodes?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rlite/rlite/pull/91#issuecomment-611985472>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABFRFQLX7LTB7FRMNTN24X3RL34OTANCNFSM4HQGOHBA>
> .
>


-- 
Vincenzo

vmaffione commented:
> I've added a working integration test.
> 
> However, it seems that if a node is both the endpoint of a flow and the Raft leader, it gets the `OpcodeFree` message twice for the same flow id, which would double free and crash the node.
> For now I've added an additional check when freeing to avoid this, as I haven't managed to track down why exactly this is happening.
> 
> Additionally, due to having to test with 10Gbps interfaces, I converted the bandwidth values in the policy to 64 bits else it would overflow.

Yes, bandwidth in bps must be 4 bits.

> What is left unchanged is
> 
>     * `dtcp_config.bandwidth` in `rlite/common.h`- I wasn't sure if this was implementation defined or spec defined so I left it alone for now

No, that's simply an issue. The spec struct is already 64 bits (`include/rina/api.h`) Please change the bandwidth value to 64 bit in `rlite/common.h`. You also need to update the kernel and bump the control API version number (`RL_API_VERSION`), since this change breaks the ABI.
> 
>     * passing policy parameter values throught `rlite-ctl`
